tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/LegacyPopTest.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Client.Cache.Items;
+using Microsoft.Identity.Client.Extensibility;
+using Microsoft.Identity.Test.Integration.net45.Infrastructure;
+using Microsoft.Identity.Test.Integration.NetFx.Infrastructure;
+using Microsoft.IdentityModel.Tokens;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Identity.Test.Integration.NetFx.HeadlessTests
+{
+    [TestClass]
+    public class LegacyPopTests
+    {
+
+        private RsaSecurityKey _currentPopKey;
+
+        [TestMethod]
+        public async Task LegacyPoPAsync()
+        {
+            IConfidentialAppSettings settings = ConfidentialAppSettings.GetSettings(Cloud.Public);
+            X509Certificate2 clientCredsCert = settings.GetCertificate();
+            _currentPopKey = CreateRsaSecurityKey();
+
+            var cca = ConfidentialClientApplicationBuilder
+                .Create(settings.ClientId)
+                .WithAuthority(settings.Authority, true)
+                .WithExperimentalFeatures(true)
+                .WithTestLogging()
+                .Build();
+
+            var result = await cca.AcquireTokenForClient(settings.AppScopes)                
+                .WithClientAssertion(
+                    (tokenEndpoint, ct) => 
+                    Task.FromResult(CreateMs10ATPOPBodyParams(settings.ClientId, tokenEndpoint, clientCredsCert, _currentPopKey, true)))
+                .WithKeyId(_currentPopKey.KeyId)
+                .ExecuteAsync().ConfigureAwait(false);
+
+            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+            MsalAccessTokenCacheItem at = (cca.AppTokenCache as ITokenCacheInternal).Accessor.GetAllAccessTokens().Single();
+            Assert.AreEqual(at.KeyId, _currentPopKey.KeyId);
+
+            result = await cca.AcquireTokenForClient(settings.AppScopes)
```

Comment:
If we do a normal acquire token call here, it will overwrite the pop token in the cache, but it's still valid, so that's why you're adding a kid to cache key?

Reply:
Yes, that's the general idea. The scenario where you request a Bearer token and a POP token for the same resource is probably quite obscure. 

The scenario where cache key includes kid is more in the lines of POP key rotation: 

## Before the changes

1. request with kid1 -> IDP
2. request with kid1 -> Cache
3. request with kid2 -> IDP
4. request with kid2 -> Cache
5. request with kid1 -> IDP

## After the cache key changes
1-4 remain the same
5. request with kid1 -> cache

When doing in-memory caching only it's not that important. But for distributed system + L2 caching it becomes a problem, if 2 machines have different POP keys.